### PR TITLE
Consumer API: Revoke a Relationship reactivation request

### PR DESCRIPTION
**Pull request by NikolaVetnic**
# Readiness checklist

-   [X] I added/updated unit tests.
-   [ ] I added/updated integration tests.
-   [X] I ensured that the PR title is good enough for the changelog.
-   [X] I labeled the PR.

# Description

Relationship reactivation can now be revoked via `PUT /Relationships/{id}/Reactivate/Revoke` endpoint, after which the status of the relationship remains `Terminated` and an appropriate `RelationshipAuditLogEntry` is created with reason `RevocationOfReactivation`. The `RelationshipStatusChangedIntegrationEvent` integration event is not triggered since the status of the Relationship doesn't change.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new feature to revoke the reactivation of relationships, enhancing control over relationship statuses.
	- Added a new HTTP endpoint to manage the revocation of relationship reactivations.

- **Enhancements**
	- Expanded support for 'Terminated' relationship status in data transfer objects.

- **Documentation**
	- Updated audit log entries to include new reasons related to relationship status changes.

- **Bug Fixes**
	- Implemented error handling for specific cases in relationship requests.

- **Tests**
	- Added comprehensive tests for new functionalities related to revoking relationship reactivations.

### CODE REVIEW

**Comment by tnotheis** [comment]
@coderabbitai review

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Synchronization/src/Synchronization.Domain/Entities/Sync/ExternalEvent.cs
Comment (NikolaVetnic) [reply Modules/Synchronization/src/Synchronization.Domain/Entities/Sync/ExternalEvent.cs:66]: Apologies, changes made.

Comment (NikolaVetnic) [reply Modules/Synchronization/src/Synchronization.Domain/Entities/Sync/ExternalEvent.cs:66]: Sure, updated. 

Changes made.

File: Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:1]: I was reluctant to remove it because I wasn't sure if it may be used by other features, in which case removing it would make merging harder. I removed it and fixed the nullability warnings.

Changes made.

File: Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationRevokedIntegrationEvent.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationRevokedIntegrationEvent.cs:8]: Changes made.

File: Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationRevokedIntegrationEvent.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationRevokedIntegrationEvent.cs:17]: Maybe I didn't understand how these function correctly. We use `CreatedAt` here to build that URL-like identifier in the base constructor. I was thinking the property is necessary in order to have the same URL on the other side in the Synchronization module.

File: Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationRevokedIntegrationEvent.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationRevokedIntegrationEvent.cs:18]: Changes made.

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/RevokeRelationshipReactivation/Handler.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/Relationships/Commands/RevokeRelationshipReactivation/Handler.cs:34]: Changes made.

File: Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationRevokedIntegrationEvent.cs
Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationRevokedIntegrationEvent.cs:17]: Ah. No, the duplicate detection happens on the actual event bus, and it just uses the id we set in the constructor. 
The "URL" as you call it isn't really a URL. It's just an id that happens to look like a URL

File: Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs
Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:1]: It's been there because Daniel Silva added it by accident. It was removed in a separate [PR](https://github.com/nmshd/backbone/pull/598) a few days ago.

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:148]: Changes made.

File: Modules/Relationships/src/Relationships.Domain/DomainErrors.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/DomainErrors.cs:53]: Hopefully I understood what you had in mind.

Changes made.

File: Modules/Relationships/test/Relationships.Application.Tests/Tests/Relationships/Commands/RevokeRelationshipReactivation/HandlerTests.cs
Comment (NikolaVetnic) [reply Modules/Relationships/test/Relationships.Application.Tests/Tests/Relationships/Commands/RevokeRelationshipReactivation/HandlerTests.cs:24]: Changes made.

File: Modules/Relationships/test/Relationships.Application.Tests/Tests/Relationships/Commands/RevokeRelationshipReactivation/HandlerTests.cs
Comment (NikolaVetnic) [reply Modules/Relationships/test/Relationships.Application.Tests/Tests/Relationships/Commands/RevokeRelationshipReactivation/HandlerTests.cs:56]: Changes made.

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:146]: Changes made.

File: Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationRevokedIntegrationEvent.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationRevokedIntegrationEvent.cs]: I know it's not really a URL, you explained it to me last week, but I was thinking it should be the same on both ends. I'll remove the `CreatedAt` property.

Changes made.

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:158]: I would rather call it "activeIdentity" than "target". "target" sounds like the target of the reactivation request.

File: Modules/Relationships/src/Relationships.Domain/DomainErrors.cs
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Domain/DomainErrors.cs:46]: See previous comment

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:164]: Reminder: we shouldn't merge this before the termination PR is merged

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:178]: Reminder: we shouldn't merge this before the reactivation PR is merged

File: Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationCompletedIntegrationEvent.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipReactivationCompletedIntegrationEvent.cs:12]: Changes made.

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:158]: Makes sense.

Changes made.

File: Modules/Relationships/src/Relationships.Domain/DomainErrors.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/DomainErrors.cs:46]: Changes made.

**Comment by NikolaVetnic** [comment]
Reminder: we shouldn't merge this before the reactivation PR is merged.

**Comment by tnotheis** [comment]
@NikolaVetnic There are conflicts in this branch.

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:158]: The method name is misleading. One might use it e.g. for the rejection of a reactivation request. But there the requirement is that the reactivation must NOT to be created by the active identity.

A better name would maybe be `EnsureRevocableReactivationRequestExistsFor`. But I'm open to other suggestions.

File: Modules/Relationships/src/Relationships.Domain/DomainErrors.cs
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Domain/DomainErrors.cs:47]: Similar as above. The error code (or method name) does not say anything about revocation. So one might use it in any scenario (accept, reject, revoke). But in the error message, you talk about "revoke".

File: Modules/Relationships/test/Relationships.Domain.Tests/TestHelpers/TestData.cs
Comment (tnotheis) [inline Modules/Relationships/test/Relationships.Domain.Tests/TestHelpers/TestData.cs:51]: Can be replaced with `CreateActiveRelationship`

File: Modules/Synchronization/src/Synchronization.Application/IntegrationEvents/Incoming/RelationshipReactivationCompleted/RelationshipReactivationCompletedIntegrationEvent.cs
Comment (tnotheis) [inline Modules/Synchronization/src/Synchronization.Application/IntegrationEvents/Incoming/RelationshipReactivationCompleted/RelationshipReactivationCompletedIntegrationEvent.cs:6]: We only need the ids for outgoing events.

```suggestion
    public RelationshipReactivationCompletedIntegrationEvent(string relationshipId, string peer)
```

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:158]: No better idea, I used yours.

Changes made.

File: Modules/Relationships/src/Relationships.Domain/DomainErrors.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/DomainErrors.cs:47]: Changes made.

File: Modules/Relationships/test/Relationships.Domain.Tests/TestHelpers/TestData.cs
Comment (NikolaVetnic) [reply Modules/Relationships/test/Relationships.Domain.Tests/TestHelpers/TestData.cs:51]: Good catch.

Changes made.

File: Modules/Synchronization/src/Synchronization.Application/IntegrationEvents/Incoming/RelationshipReactivationCompleted/RelationshipReactivationCompletedIntegrationEvent.cs
Comment (NikolaVetnic) [reply Modules/Synchronization/src/Synchronization.Application/IntegrationEvents/Incoming/RelationshipReactivationCompleted/RelationshipReactivationCompletedIntegrationEvent.cs:6]: Changes made.

File: Modules/Synchronization/src/Synchronization.Application/DomainEvents/Incoming/RelationshipReactivationCompleted/RelationshipReactivationCompletedIntegrationEventHandler.cs
Comment (tnotheis) [inline Modules/Synchronization/src/Synchronization.Application/DomainEvents/Incoming/RelationshipReactivationCompleted/RelationshipReactivationCompletedIntegrationEventHandler.cs:42]:
```suggestion
            _logger.LogError(ex, "An error occured while processing an domain event.");
```

Can you check the other event handlers as well and fix it there if necessary?

File: Modules/Synchronization/src/Synchronization.Application/Synchronization.Application.csproj
Comment (tnotheis) [inline Modules/Synchronization/src/Synchronization.Application/Synchronization.Application.csproj:21]: Remove this

**Review by tnotheis — CHANGES_REQUESTED** [review]
Did you test everything manually? "everything" means: calling the new endpoint, checking the returned result, the database content after the request, and whether and external event was created

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs]: Changes made.

File: Modules/Relationships/src/Relationships.Domain/DomainErrors.cs
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/DomainErrors.cs:47]: Changes made.

File: Modules/Relationships/test/Relationships.Application.Tests/TestHelpers/TestData.cs
Comment (NikolaVetnic) [reply Modules/Relationships/test/Relationships.Application.Tests/TestHelpers/TestData.cs:50]: 1. I agree the variable names are a bit ambiguous, but the code is not - it's clear that `activeIdentity` is the one creating a relationship, terminating it and then requesting reactivation. I didn't think all that can or should be cramped into a long variable name, and I am not sure if there is a nice short variant that says it all.
2. We don't, I updated the method.
3. I added the `CreateTerminatedRelationship` that accepts the `from` and `to` parameters and used that.
4. Ok, I did, now my method looks just like the equivalent one over there, with the exception of static read only fields `IDENTITY_1`, `DEVICE_1`, and so on. If you'd like me to introduce that as well, then I think update of all tests would be in order. Should I do that?

File: Modules/Relationships/test/Relationships.Domain.Tests/Tests/Aggregates/Relationships/RevokeRelationshipReactivationTests.cs
Comment (NikolaVetnic) [reply Modules/Relationships/test/Relationships.Domain.Tests/Tests/Aggregates/Relationships/RevokeRelationshipReactivationTests.cs:13]: Changes made.

File: Modules/Synchronization/src/Synchronization.Application/DomainEvents/Incoming/RelationshipReactivationCompleted/RelationshipReactivationCompletedIntegrationEventHandler.cs
Comment (NikolaVetnic) [reply Modules/Synchronization/src/Synchronization.Application/DomainEvents/Incoming/RelationshipReactivationCompleted/RelationshipReactivationCompletedIntegrationEventHandler.cs:42]: Changed all of the event handlers.

Changes made.

File: Modules/Synchronization/src/Synchronization.Application/Synchronization.Application.csproj
Comment (NikolaVetnic) [reply Modules/Synchronization/src/Synchronization.Application/Synchronization.Application.csproj:21]: Changes made.

**Review by tnotheis — APPROVED** [review]

**Pull request merged**
